src/glue.c

```diff
@@ -214,6 +217,135 @@ PyObject *mg_path_to_py_path(const mg_path *path) {
+  PyDateTime_IMPORT;
+  PyObject *delta = PyDelta_FromDSU(days, sec, micros);
+  if (!delta) {
+    PyErr_Print();
+  }
+  return delta;
+}
+
+PyObject *mg_date_to_py_date(const mg_date *date) {
+  PyDateTime_IMPORT;
+  SCOPED_CLEANUP PyObject *unix_epoch = make_py_date(1970, 1, 1);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(unix_epoch, NULL)
+  SCOPED_CLEANUP PyObject *date_as_delta =
+      make_py_delta(mg_date_days(date), 0, 0);
+  SCOPED_CLEANUP PyObject *method_name = PyUnicode_FromString("__add__");
+  PyObject *result_date =
+      PyObject_CallMethodObjArgs(unix_epoch, method_name, date_as_delta, NULL);
+  if (!result_date) {
+    PyErr_Print();
+  }
+  return result_date;
+}
+
+PyObject *mg_local_time_to_py_time(const mg_local_time *lt) {
+  PyDateTime_IMPORT;
+  SCOPED_CLEANUP PyObject *unix_epoch =
+      make_py_datetime(1970, 1, 1, 0, 0, 0, 0);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(unix_epoch, NULL)
+  SCOPED_CLEANUP PyObject *numerator =
+      PyLong_FromLong(mg_local_time_nanoseconds(lt));
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(numerator, NULL)
+  SCOPED_CLEANUP PyObject *denominator = PyLong_FromLong(1000000000);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(denominator, NULL)
+  SCOPED_CLEANUP PyObject *div_result =
+      PyNumber_TrueDivide(numerator, denominator);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(div_result, NULL)
+  SCOPED_CLEANUP PyObject *method_name =
+      PyUnicode_FromString("utcfromtimestamp");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(method_name, NULL)
+  SCOPED_CLEANUP PyObject *result =
+      PyObject_CallMethodObjArgs(unix_epoch, method_name, div_result, NULL);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(result, NULL);
+  SCOPED_CLEANUP PyObject *h = PyObject_GetAttrString(result, "hour");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(h, NULL)
+  SCOPED_CLEANUP PyObject *m = PyObject_GetAttrString(result, "minute");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(m, NULL)
+  SCOPED_CLEANUP PyObject *s = PyObject_GetAttrString(result, "second");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(s, NULL)
+  SCOPED_CLEANUP PyObject *mi = PyObject_GetAttrString(result, "microsecond");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(mi, NULL)
+  return make_py_time(PyLong_AsLong(h), PyLong_AsLong(m), PyLong_AsLong(s),
+                      PyLong_AsLong(mi));
+}
+
+PyObject *mg_local_date_time_to_py_datetime(const mg_local_date_time *ldt) {
+  PyDateTime_IMPORT;
+  SCOPED_CLEANUP PyObject *unix_epoch =
+      PyDateTime_FromDateAndTime(1970, 1, 1, 0, 0, 0, 0);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(unix_epoch, NULL)
+  SCOPED_CLEANUP PyObject *seconds =
+      PyLong_FromLong(mg_local_date_time_seconds(ldt));
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(seconds, NULL)
+  SCOPED_CLEANUP PyObject *method_name =
+      PyUnicode_FromString("utcfromtimestamp");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(method_name, NULL)
+  SCOPED_CLEANUP PyObject *result =
+      PyObject_CallMethodObjArgs(unix_epoch, method_name, seconds, NULL);
```

Comment:
If you use [`PyObject_CallMethod`](https://docs.python.org/3/c-api/call.html#c.PyObject_CallMethod) then you don't need to create a `PyUnicode`.

The same true for all other occurrence of `PyObject_CallMethodObjArgs`.

Reply:
Yes but it has other downsides that are describe here: `https://docs.python.org/3/c-api/call.html#c.PyObject_CallMethod` 

1. For the C arguments I have to use `The C arguments are described by a Py_BuildValue()`
2. `Note that if you only pass PyObject * args, PyObject_CallMethodObjArgs() is a faster alternative.`

Reply:
1. That's true, but I think that's fine as this is how Python API works.
2. That might be true, but when you create the `PyUnicode` it will copy all the bytes, allocate etc.

What if we define these values for global object for this file (or globally in the whole module)? Then we can use `PyObject_CallMethodObjArgs` without copying the same bytes over and over again. Or maybe create a static variable inside the function, though I am not sure if that is possible in C or not.

Reply:
1. Yes but if I have an alternative why would I not choose that? They are both API functions.
2. Wouldn't it be the same case with args? You, still need to allocate on these cases as well ?
I think you can use static variables in a function but this creates a new problem. Who owns the lifetime of the string? Unfortunately, `SCOPED_CLEANUP` works only for  automatic storage duration.

We decided offline to check alternatives and we will update this post :)